Better Errors: team policy inner loop args

PyKokkos team parallel dispatches *do not* support 
argument parsing. For example, consider https://github.com/kokkos/pykokkos/blob/main/examples/kokkos-tutorials/standalone/team_policy.py.
If we change the workunit to

```
@pk.workunit
def yAx(team_member, acc, cols, y_view, x_view, A_view):
    j: int = team_member.league_rank()

    def inner_reduce(i: int, inner_acc: pk.Acc[float], c: float):
        inner_acc += A_view[j][i] * x_view[i] + c 

    temp2: float = pk.parallel_reduce(
        pk.TeamThreadRange(team_member, cols), inner_reduce, c=0.5
    )   

    if team_member.team_rank() == 0:
        acc += y_view[j] * temp2
```

that is, if we pass in a key-word argument to `inner_reduce`, the translation fails with

```
/work/09661/gkk345/vista/pykokkos/pk_cpp/examples/kokkos-tutorials/standalone/team_policy/team_policy_yAx/types_a5d3098b014ebac4e68e6114dd8b833e/Cuda/../functor.hpp(29): error: identifier "inner_acc" is undefined
            inner_acc += ((((A_view(j, i)) * (x_view(i)))) + (c)); 
            ^

/work/09661/gkk345/vista/pykokkos/pk_cpp/examples/kokkos-tutorials/standalone/team_policy/team_policy_yAx/types_a5d3098b014ebac4e68e6114dd8b833e/Cuda/../functor.hpp(29): error: identifier "c" is undefined
            inner_acc += ((((A_view(j, i)) * (x_view(i)))) + (c)); 
                                                              ^

2 errors detected in the compilation of "/work/09661/gkk345/vista/pykokkos/pk_cpp/examples/kokkos-tutorials/standalone/team_policy/team_policy_yAx/types_a5d3098b014ebac4e68e6114dd8b833e/Cuda/bindings.cpp".
gmake[2]: *** [CMakeFiles/kernel.cpython-313-aarch64-linux-gnu.dir/build.make:79: CMakeFiles/kernel.cpython-313-aarch64-linux-gnu.dir/bindings.cpp.o] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:87: CMakeFiles/kernel.cpython-313-aarch64-linux-gnu.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2
```

This is orthogonal to how outer loop argument parsing works, and it should be fixed with either better error messages or functionality support.

--- a/pykokkos/core/translators/symbols_pass.py
+++ b/pykokkos/core/translators/symbols_pass.py
@@ -128,6 +128,19 @@ class SymbolsPass:
             elif isinstance(node, ast.FunctionDef):
                 if self.is_nested_call(node):
                     symbols.add(node.name)
+            elif isinstance(node, ast.Call):
+                # kwargs passed to a nested parallel_reduce/parallel_scan become
+                # local auto variables at the call site (captured by [&] in the
+                # inner lambda), so treat their names as locally defined symbols.
+                func_name: str = ""
+                if isinstance(node.func, ast.Attribute):
+                    func_name = node.func.attr
+                elif isinstance(node.func, ast.Name):
+                    func_name = node.func.id
+                if func_name in ("parallel_reduce", "parallel_scan", "parallel_for"):
+                    for kw in node.keywords:
+                        if kw.arg is not None:
+                            symbols.add(kw.arg)
 
         return symbols
 

--- a/pykokkos/core/visitors/workunit_visitor.py
+++ b/pykokkos/core/visitors/workunit_visitor.py
@@ -168,7 +168,17 @@ class WorkunitVisitor(PyKokkosVisitor):
 
                 callstmt = cppast.CallStmt(call)
 
-                return cppast.CompoundStmt([declstmt, callstmt])
+                # Declare kwargs as local auto variables so the inner lambda's
+                # [&] capture can pick them up.
+                kwarg_stmts: List[cppast.DeclStmt] = []
+                for kw in node.value.keywords:
+                    kw_type = cppast.PrimitiveType("auto")
+                    kw_name = cppast.DeclRefExpr(kw.arg)
+                    kw_value = self.visit(kw.value)
+                    kw_decl = cppast.VarDecl(kw_type, kw_name, kw_value)
+                    kwarg_stmts.append(cppast.DeclStmt(kw_decl))
+
+                return cppast.CompoundStmt(kwarg_stmts + [declstmt, callstmt])
 
             if function_name.startswith("ScratchView"):
                 cpp_view_type: str = self.get_scratch_view_type(node.annotation)
@@ -185,6 +195,21 @@ class WorkunitVisitor(PyKokkosVisitor):
                 return cppast.DeclStmt(view_decl)
 
         return super().visit_AnnAssign(node)
+
+    def visit_Expr(self, node: ast.Expr) -> cppast.Stmt:
+        if isinstance(node.value, ast.Call) and node.value.keywords:
+            func_name: str = visitors_util.get_node_name(node.value.func)
+            if func_name in ("parallel_for", "parallel_scan"):
+                kwarg_stmts: List[cppast.DeclStmt] = []
+                for kw in node.value.keywords:
+                    kw_type = cppast.PrimitiveType("auto")
+                    kw_name = cppast.DeclRefExpr(kw.arg)
+                    kw_value = self.visit(kw.value)
+                    kw_decl = cppast.VarDecl(kw_type, kw_name, kw_value)
+                    kwarg_stmts.append(cppast.DeclStmt(kw_decl))
+                call = self.visit(node.value)
+                return cppast.CompoundStmt(kwarg_stmts + [cppast.CallStmt(call)])
+        return super().visit_Expr(node)
 
     def visit_arguments(self, node: ast.arguments) -> List[cppast.ParmVarDecl]:
         args: List[ast.arg] = node.args

--- a/tests/test_hierarchical.py
+++ b/tests/test_hierarchical.py
@@ -92,6 +92,43 @@ class HierarchicalTestFunctor:
             self.for_view[j] = temp
 
     @pk.workunit
+    def yAx_scaled(self, team_member: pk.TeamMember, acc: pk.Acc[pk.double]) -> None:
+        j: int = team_member.league_rank()
+
+        def inner_reduce(i: int, inner_acc: pk.Acc[pk.double]):
+            inner_acc += self.A[j][i] * self.x[i] * scale
+
+        temp2: float = pk.parallel_reduce(
+            pk.TeamThreadRange(team_member, self.M), inner_reduce, scale=2.0
+        )
+
+        if team_member.team_rank() == 0:
+            acc += self.y[j] * temp2
+
+    @pk.workunit
+    def fill_offset(self, team_member: pk.TeamMember) -> None:
+        j: int = team_member.league_rank()
+
+        def inner_for(i: int):
+            self.A[j][i] = offset
+
+        pk.parallel_for(pk.TeamThreadRange(team_member, self.M), inner_for, offset=42)
+
+    @pk.workunit
+    def scan_scaled(self, team_member: pk.TeamMember, acc: pk.Acc[pk.double]) -> None:
+        j: int = team_member.league_rank()
+
+        def inner_scan(i: int, scan_acc: pk.Acc[pk.double], is_final: bool):
+            scan_acc += scale
+
+        result: float = pk.parallel_scan(
+            pk.TeamThreadRange(team_member, self.M), inner_scan, scale=3
+        )
+
+        if team_member.team_rank() == 0:
+            acc += result
+
+    @pk.workunit
     def yAx_vector(self, team_member: pk.TeamMember, acc: pk.Acc[pk.double]) -> None:
         e: int = team_member.league_rank()
 
@@ -194,6 +231,42 @@ class TestHierarchical(unittest.TestCase):
         for i in range(self.N):
             result: int = self.functor.for_view[i]
             self.assertEqual(expected_result, result)
+
+    def test_yAx_scaled(self):
+        # Regression test for issue #409: kwargs passed to nested parallel_reduce
+        # inside a team policy workunit were silently dropped, causing C++ compile
+        # errors ("identifier 'scale' is undefined").
+        scale: float = 2.0
+        expected_result: float = 0
+        for j in range(self.N):
+            temp2: float = 0
+            for i in range(self.M):
+                temp2 += self.A[j][i] * self.x[i] * scale
+            expected_result += self.y[j] * temp2
+
+        result: float = pk.parallel_reduce(
+            pk.TeamPolicy(self.execution_space, self.N, pk.AUTO), self.functor.yAx_scaled
+        )
+
+        self.assertEqual(expected_result, result)
+
+    def test_fill_offset(self):
+        # Regression test for issue #409: kwargs passed to nested parallel_for
+        pk.parallel_for(
+            pk.TeamPolicy(self.execution_space, self.N, pk.AUTO), self.functor.fill_offset
+        )
+        for j in range(self.N):
+            for i in range(self.M):
+                self.assertEqual(self.functor.A[j][i], 42)
+
+    def test_scan_scaled(self):
+        # Regression test for issue #409: kwargs passed to nested parallel_scan
+        # Each team scans M elements each contributing `scale=3`, giving total 3*M per team.
+        expected_result: float = self.N * self.M * 3
+        result: float = pk.parallel_reduce(
+            pk.TeamPolicy(self.execution_space, self.N, pk.AUTO), self.functor.scan_scaled
+        )
+        self.assertEqual(expected_result, result)
 
     def test_yAx_vector(self):
         expected_result: float = 0

--- a/tests/test_hierarchical.py
+++ b/tests/test_hierarchical.py
@@ -242,14 +242,16 @@ class TestHierarchical(unittest.TestCase):
             expected_result += self.y[j] * temp2
 
         result: float = pk.parallel_reduce(
-            pk.TeamPolicy(self.execution_space, self.N, pk.AUTO), self.functor.yAx_scaled
+            pk.TeamPolicy(self.execution_space, self.N, pk.AUTO),
+            self.functor.yAx_scaled,
         )
 
         self.assertEqual(expected_result, result)
 
     def test_fill_offset(self):
         pk.parallel_for(
-            pk.TeamPolicy(self.execution_space, self.N, pk.AUTO), self.functor.fill_offset
+            pk.TeamPolicy(self.execution_space, self.N, pk.AUTO),
+            self.functor.fill_offset,
         )
         for j in range(self.N):
             for i in range(self.M):
@@ -258,7 +260,8 @@ class TestHierarchical(unittest.TestCase):
     def test_scan_scaled(self):
         expected_result: float = self.N * self.M * 3
         result: float = pk.parallel_reduce(
-            pk.TeamPolicy(self.execution_space, self.N, pk.AUTO), self.functor.scan_scaled
+            pk.TeamPolicy(self.execution_space, self.N, pk.AUTO),
+            self.functor.scan_scaled,
         )
         self.assertEqual(expected_result, result)
 

--- a/tests/test_hierarchical.py
+++ b/tests/test_hierarchical.py
@@ -233,9 +233,6 @@ class TestHierarchical(unittest.TestCase):
             self.assertEqual(expected_result, result)
 
     def test_yAx_scaled(self):
-        # Regression test for issue #409: kwargs passed to nested parallel_reduce
-        # inside a team policy workunit were silently dropped, causing C++ compile
-        # errors ("identifier 'scale' is undefined").
         scale: float = 2.0
         expected_result: float = 0
         for j in range(self.N):
@@ -251,7 +248,6 @@ class TestHierarchical(unittest.TestCase):
         self.assertEqual(expected_result, result)
 
     def test_fill_offset(self):
-        # Regression test for issue #409: kwargs passed to nested parallel_for
         pk.parallel_for(
             pk.TeamPolicy(self.execution_space, self.N, pk.AUTO), self.functor.fill_offset
         )
@@ -260,8 +256,6 @@ class TestHierarchical(unittest.TestCase):
                 self.assertEqual(self.functor.A[j][i], 42)
 
     def test_scan_scaled(self):
-        # Regression test for issue #409: kwargs passed to nested parallel_scan
-        # Each team scans M elements each contributing `scale=3`, giving total 3*M per team.
         expected_result: float = self.N * self.M * 3
         result: float = pk.parallel_reduce(
             pk.TeamPolicy(self.execution_space, self.N, pk.AUTO), self.functor.scan_scaled

--- a/tests/test_hierarchical.py
+++ b/tests/test_hierarchical.py
@@ -1,5 +1,6 @@
 import unittest
 
+import pytest
 import pykokkos as pk
 
 
@@ -153,6 +154,38 @@ class HierarchicalTestFunctor:
         pk.single(pk.PerTeam(team_member), single_closure)
 
 
+@pk.functor
+class ShadowTestFunctor:
+    def __init__(self, N: int, M: int, value: int):
+        self.N: int = N
+        self.M: int = M
+        self.y: pk.View1D[pk.int32] = pk.View([N], pk.int32)
+        self.x: pk.View1D[pk.int32] = pk.View([M], pk.int32)
+        self.A: pk.View2D[pk.int32] = pk.View([N, M], pk.int32)
+        for i in range(N):
+            self.y[i] = value
+        for i in range(M):
+            self.x[i] = value
+        for j in range(N):
+            for i in range(M):
+                self.A[j][i] = value
+
+    @pk.workunit
+    def yAx_shadow(self, team_member: pk.TeamMember, acc: pk.Acc[pk.double]) -> None:
+        j: int = team_member.league_rank()
+        scale: float = 1.0
+
+        def inner_reduce(i: int, inner_acc: pk.Acc[pk.double]):
+            inner_acc += self.A[j][i] * self.x[i] * scale
+
+        temp2: float = pk.parallel_reduce(
+            pk.TeamThreadRange(team_member, self.M), inner_reduce, scale=2.0
+        )
+
+        if team_member.team_rank() == 0:
+            acc += self.y[j] * temp2
+
+
 class TestHierarchical(unittest.TestCase):
     def setUp(self):
         self.N: int = 64
@@ -247,6 +280,20 @@ class TestHierarchical(unittest.TestCase):
         )
 
         self.assertEqual(expected_result, result)
+
+    @pytest.mark.xfail(
+        reason="kwarg name shadowing an existing local variable emits a C++ "
+        "redefinition in the same scope; variable shadowing is not yet handled"
+    )
+    def test_yAx_shadow(self):
+        # A kwarg whose name matches an existing local variable ('scale') should
+        # A kwarg whose name matches an existing local variable ('scale') should
+        # eventually be handled via scoping, but currently produces a C++ compile error.
+        shadow_functor = ShadowTestFunctor(self.N, self.M, self.value)
+        pk.parallel_reduce(
+            pk.TeamPolicy(self.execution_space, self.N, pk.AUTO),
+            shadow_functor.yAx_shadow,
+        )
 
     def test_fill_offset(self):
         pk.parallel_for(

